doc/protocol.md

```diff
@@ -829,6 +829,119 @@ or pipe is done.
+ * `close-on-error` (optional, default: true): if the channel should close
+   (with `problem`) on failure to `open()` the named `path`
+ * `watch` (optional, default: false): if the channel should watch for changes
+ * `follow` (default: true): if the trailing component of `path` is a symbolic
+   link, whether we should follow it.  `no-follow` mode is not supported for
+   `watch` (but is also mostly unnecessary).
+
+A channel is created with an associated absolute pathname ('path').  A path is
+a string that you pass to the `open()` syscall.  The object returned by that
+call is described by the (JSON Document) value of the channel.
+
+Strictly speaking, the path is specified to the channel as a unicode string,
+but paths on the filesystem are binary strings.  This mapping is handled using
+the usual Python convention: filenames are stored in utf8 and encoded and
+decoded with `errors='surrogateescape'`, which allows for handling of filenames
+which are not utf8.
+
+Conceptually, the channel represents the value of a JSON object describing what
+is found at the path, or the reason for an error.  Each message on the channel
+is a JSON Merge Patch (RFC7396) which updates or replaces this object.
+
+The top-level keys in the object are:
+
+ - `info`: the info about the reported path (described below)
+ - `problem`: a "problem code" like `not-found` or `not-directory`
+ - `message`: a user-readable (`strerror`) string describing the error
+ - `partial`: `true` if the data is in a "partial" state.  This is used when
+   sending updates in multiple pieces.
+
+If `close-on-error` is set, `problem` and `message` will never be sent, but the
+channel will be closed with the `problem` code set to indicate the error.
+
+Otherwise, the value of the `info` attribute is a JSON object describing the
+attributes of the file.  Any attribute which cannot be read (eg: due to
+permissions problems) is simply not reported.
+
+ * `type`: a string: `file`, `directory`, `link`, or `special`
+ * `mode`: an integer representing the "file permission" bits.  As per JSON,
+   this is transmitted as a decimal value, but it is meant to be interpreted in
+   octal, in the usual way.
+ * `uid`: an integer, the uid of the file owner (`st_uid`)
+ * `owner`: a string, or an integer if the lookup failed
+ * `gid`: an integer, the gid of the file group (`st_gid`)
+ * `group`: a string, or an integer if the lookup failed
+ * `size`: an integer, the (apparent) size of the file (`st_size`)
+ * `modified`: a float, the mtime of the file (`st_mtim`)
+ * `tag`: a value type: the current 'transaction tag' of the file, with the
+   same meaning as elsewhere in this document.  Ideally: this changes if the
+   content of the file changes.  This is stable between different channels.
+ * `entries`: for directories: the contents of the directory (see below)
+ * `content`: for regular text files, the content (if less than 4096 bytes)
```

Comment:
That feels a bit arbitrary and too rigid. Shouldn't the caller decide via `attrs` whether it's sensible to directly read the files (if they need it in a race-free manner) or not (even if they are small)? Cf.  https://cockpit-project.org/guide/latest/cockpit-file.html#cockpit-file-notify . For safety I'd rather see a `max_read_size` option like fsread1 has.

Reply:
The caller does decide via `attrs` to read file content: by requesting the `content` attr.  It will never be reported if not requested.

I considered adding `max-read-size` but figured we could wait-and-see if it's actually needed.  I actually have another plan here, anyway: I do want to support large files, but I want to do it via opening an external channel somehow (which needs some thinking — we currently have no concept of "push channels").

That's why I decided to leave this one hanging for the time being...

Reply:
Hmm, so perhaps this PR should drop "content" altogether for now?

Reply:
It's gone now. This is a bit tricky, so let's introduce it in a follow-up if/when we get a consumer.